### PR TITLE
Python 3.10

### PR DESCRIPTION
**Pull request by squeaky-pl**
Supersedes: https://github.com/closeio/sync-engine/pull/439

### CODE REVIEW

File: inbox/util/html.py [hunk 1]
```diff
@@ -68,7 +68,14 @@ def strip_tags(html: str) -> str:
...
     """
     s = HTMLTagStripper()
-    s.feed(html)
+
+    try:
+        s.feed(html)
+    except AssertionError as e:
+        if e.args[0].startswith("unknown status keyword"):
+            raise HTMLParseError(*e.args) from e
+        raise
```
Comment (wojcikstefan) [inline inbox/util/html.py:77]: Curious, why was this change needed?

Comment (squeaky-pl) [reply inbox/util/html.py:77]: The python parser in Python 3.10 started throwing AssertionErrors on some invalid HTML, there were previously HTMLParserError. I think it's opinionated weather it should be AssertionErrors, especially when treating with user supplied data (which emails are). We actually have a test for those, meaning that at some point Nylas people hit this kind of HTML that would now trigger AssertionError.

Comment (squeaky-pl) [reply inbox/util/html.py:77]: @wojcikstefan should I add clarification comment maybe?

Comment (wojcikstefan) [reply inbox/util/html.py:77]: Thanks for explaining, makes sense 👍 Yea, I think it's worth adding a comment – it might not be clear at a glance why we don't work directly with the `AssertionError` further up the stack.

Are there other invalid HTML errors that will still raise an `AssertionError`? Should we re-raise **all** `s.feed(html)` assertion errors as HTML parser errors? Or does `s.feed` do more than just parse?

Comment (squeaky-pl) [reply inbox/util/html.py:77]: Additionally if those would become AssertionErrors we would not catch them [here](https://github.com/closeio/sync-engine/blob/4f192bca8af9c8f1b91e25a0809cf14954a3d64c/inbox/models/message.py#L656-L665)

```
    def calculate_html_snippet(self, text: str) -> str:
        try:
            text = strip_tags(text)
        except HTMLParseError:
            log.error(
                "error stripping tags", message_nylas_uid=self.nylas_uid, exc_info=True
            )
            text = ""

        return self.calculate_plaintext_snippet(text)
```

meaning that we would discard the entire email, instead of discarding just the snippet.

Comment (wojcikstefan) [reply inbox/util/html.py:77]: We technically could replace `except HTMLParserError` with `except AssertionError` though, right? Or are there other `AssertionErrors` being thrown that we should definitely NOT catch?

Comment (squeaky-pl) [reply inbox/util/html.py:77]: @wojcikstefan I did a deep dive, found the change in CPython that causes this:

https://github.com/python/cpython/commit/e34bbfd61f405eef89e8aa50672b0b25022de320

As you can see they have removed the [error callback](https://github.com/python/cpython/commit/e34bbfd61f405eef89e8aa50672b0b25022de320#diff-7304624ffcc8be1268e5f86ebf1723034477694ce33456dc330f9c03f2d1e84fL32-L35), we were using the error callback to raise [our exception](https://github.com/closeio/sync-engine/blob/3fc0382512bba63e1a9fe551910c2a069b20615a/inbox/util/html.py#L46-L48). But that method was [long deprecated](https://github.com/python/cpython/issues/76025). I still think the choice to raise AssertionError is the wrong choice of CPython developers. So I agree we should catch all the AssertionErrors.

There are some AssertionErrors that are programming errors in the source of the parser, like [this one](https://github.com/python/cpython/blob/ffee63f34445412322a7afc2535731be221cabba/Lib/html/parser.py#L382C7-L382C56), but they should not happen.

Thanks for catching this.  Will follow with a fix.

Comment (squeaky-pl) [reply inbox/util/html.py]: @wojcikstefan Pushed a fix in [`4ae8c49` (#959)](https://github.com/closeio/sync-engine/pull/959/commits/4ae8c4975a445914b2ebc79587b14c351356eee5). Assigning you as a reviewer as well because you spent a lot of time looking at this already.

**Review by wojcikstefan — APPROVED** [review]
Looks good! Do you think it's worth rolling out partially first to confirm we didn't miss anything?

**Comment by squeaky-pl** [comment]
@wojcikstefan Yes, I will definitely canary this.
